### ISSUE TITLE
Introduce Cost Model for Query Optimization

### ISSUE THREAD

**Issue by mkleen** [crate/crate#14146]
We have an iterative optimizer in place, now the next step is to introduce a cost model. Let's have a look how other system handle the cost model:

- In Apache Calcite, the cost is modelled as a scalar representing the number of rows being processed.

- In Apache Spark the cost is a vector of the number of rows and the number of bytes being processed. The vector is converted into a scalar value during comparison.

- In Presto the cost is a vector of estimated CPU, memory, and network usage. The vector is also converted into a scalar value during comparison.

- In CockroachDB, the cost is an abstract 64-bit floating-point scalar value. 

https://github.com/crate/crate/pull/14119 already introduced number of rows and number of bytes being processed ready to use for the Optimizer.

**Comment by mfussenegger** [comment]
Note that in the join order benchmark (http://www.vldb.org/pvldb/vol9/p204-leis.pdf, section 5.4) they mention that the results with a pretty simple cost model function were pretty good, and, the impact of the quality of the cost model is small compared to impact of cardinality estimates and join ordering algorithm. So I think to start with we should go with a simple variant.

**Comment by theartpiece** [comment]
Hi

I am Chitrank, and I am currently pursuing MS in CS at UT Austin.

As part of my course project, I have to work on addressing a specific issue or bug within a defined timeframe, with the goal of creating a Pull Request (PR) to contribute to the project's development. Given the significance of this project, I am seeking assistance and guidance from experienced contributors such as yourself.

I have identified a set of issues that align with my project's objectives, and I believe that your expertise and insights would be immensely valuable in helping me understand and resolve these challenges effectively.

Kind regards,

Chitrank

**Comment by mfussenegger** [comment]
Note that we ended up disabling `EquiJoinToLookupJoin` by default due to the increased memory usage. This to me implies that the cost model must include the required memory in some form.

**Comment by mkleen** [comment]
Yes i agree. The write up tries to explain the difference between fully cost-based approach like cockroachdb or hive against a partial cost-approach like presto or duckdb. I was focused on the optimizer approach not so much on the cost model itself.

**Comment by henrikingo** [comment]
Just reading up on things... @mkleen is the goal to release this work in 5.9, but put it behind a flag so users can choose the old or new optimizer? And which one would be on by default?

**Comment by matriv** [comment]
Cost-based optimizer was not planned at all for 5.9

**Comment by henrikingo** [comment]
Ah right, it is back in Candidates. Thanks @matriv .
